lms/djangoapps/shoppingcart/views.py

```diff
@@ -1022,3 +1031,118 @@ def csv_report(request):
+            course_id, user=request.user,
+            ip_address=get_ip(request),
+            url=request.path
+        )
+        if redirect_url:
+            return Response(
+                status=status.HTTP_403_FORBIDDEN,
+                data={
+                    "message": (
+                        u"Users from this location cannot access the course '{course_id}'."
+                    ).format(course_id=course_id),
+                    "user_message_url": redirect_url
+                }
+            )
+
+        # Get the course enrollment details from the enrollment API, then determine if the
+        # honor mode has a SKU associated with it. This will indicate that there is an
+        # associated product in the catalog, and we can use the e-commerce service to
+        # purchase the seat.
+        course_details = enrollment_api.get_course_enrollment_details(unicode(course_id))
+        sku = None
+        for mode in course_details['course_modes']:
+            if mode['slug'] == 'honor' and mode['sku']:
+                sku = mode['sku']
+                break
+
+        if sku is not None:
+            return self._call_ecommerce_purchase_endpoint(user, sku)
+        else:
+            try:
+                return Response(enrollment_api.add_enrollment(user, unicode(course_id)))
+            except CourseEnrollmentError as error:
+                return Response(
+                    status=status.HTTP_400_BAD_REQUEST,
+                    data={
+                        "message": u"Error enrolling for course '{course_id}'.".format(course_id=course_id),
+                        "course_details": error.data
+                    })
+
+    def _call_ecommerce_purchase_endpoint(self, username, sku):
+        """ Make a request to the E-Commerce Service purchase endpoint
+
+        Makes an HTTP request to the E-Commerce Service's purchase endpoint for a User and SKU.
+
+        Args:
+            username (string): Make a purchase on behalf of the specified user.
+            sku (string): The SKU for the product being purchased.
+
+        Returns:
+            response (Response): Return the response from the E-Commerce Service endpoint.
+
+        """
+        log.info("Attempting to purchase product with SKU [%s]", sku)
+        purchase_api_url = getattr(settings, 'PURCHASE_API_URL', None)
+        api_key = getattr(settings, 'EDX_API_KEY', None)
+        if not purchase_api_url or not api_key:
+            error_msg = "PURCHASE_API_URL and EDX_API_KEY must be set to purchase a course seat."
+            log.error(error_msg)
+            return Response(
+                status=status.HTTP_500_INTERNAL_SERVER_ERROR,
+                data={"message": error_msg}
+            )
+        data = {'user': username, 'sku': sku,}
+        headers = {'Content-Type': 'application/json', 'X-Edx-Api-Key': api_key,}
+        response = requests.post(purchase_api_url, data=json.dumps(data), headers=headers, timeout=5)
```

Comment:
Unless I'm misunderstanding something, a JWT seems like it would work well here for communicating with the e-commerce service. `data` could be augmented with extra user claims (e.g., email address) as well as an [expiration time claim](https://github.com/jpadilla/pyjwt/#expiration-time-claim), then encoded as the payload of a JWT as follows:
```python
>>> token = jwt.encode(data, 'top-secret-key')
>>> token
'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJza3UiOiJBLVNLVSJ9.SbPiPT5t-9g6DLjXSX12nS_swLp7_m417_p4B-dNWYk'
>>> response = requests.post(purchase_api_url, data=json.dumps({'token': token}), timeout=5)
```
If we were to go this route, it doesn't seem like we'd need to include the `X-Edx-Api-Key` header in the request.